### DISCUSSION TITLE
Auto-generate Slug for Each Locale

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8116] [Ideas]
> Created by: Chris Comfort (chris@identity.utexas.edu) on 2015/03/31 22:13:42 +0000
> Votes at time of UserVoice import: 35

-----------------------------------

As we all know, Craft auto-generates a slug for an entry's initial title. However, after switching to a different locale and adding its locale-specific (i.e. translated) title, the slug does not auto-generate—instead, the original slug persists through each locale.

My initial expectation was that each locale would auto-generate a new slug based on its title. This seems to fall somewhere between a bug and a feature request, so I'm posting here for everyone's thoughts.

Cheers.

**Comment by angrybrad** [comment]
> Posted by Eric (echantigny@carbure.co) on 2016/04/04 20:56:47 +0000
-----------------------------------

I fully agree with this request.  I just migrated a 500 pages site, in 2 languages, and it would&#39;ve been great to not have to deal with the slugs at all.  It&#39;s easy enough to open each page, delete the slug and save, but 500 times is a lot.

**Comment by angrybrad** [comment]
> Posted by Mats Mikkel Rummelhoff (mail@mmikkel.no) on 2015/03/09 00:12:31 +0000
-----------------------------------

I concur that the current behaviour – where you need to manually edit titles and slugs for all entries in all locales, _after_ the entry has been created – is not client proof at all.

This idea is seriously half-baked, but: wouldn&#39;t it be awesome if it were possible to enter a translated title for *all enabled locales* when you create a new entry? 

Say you have content in English, Norwegian, Danish and Swedish. When you create a new entry – regardless of the locale that entry is created in – you have the option to set the translated titles (and slugs) right there. No need to manually edit that stuff for every entry in every locale, _after_ the entry has first been created. If you don&#39;t enter a title or a slug for a specific locale, that data gets copied over from the original entry (just like it works now – the slug should be auto-generated from the localized title though, if you only translate the titles).

Come to think of it, the option to edit titles and slugs for all locales should probably be available at all times, in all locales – not only when the entry is first created.

I&#39;m not sure how the UI for titles would look, but here&#39;s a quick mockup of how it could work for slugs: Here&#39;s a quick mockup of how it could look for slugs: http://cly.mmikkel.no/image/2N0f1H2H0v42

**Comment by angrybrad** [comment]
> Posted by Christian (seelbach@gmx.de) on 2015/03/09 00:49:13 +0000
-----------------------------------

@Brandon: what if the other locales would get everything copied over except the slug on the first save? This is probably only possible if they would also stay disabled until they get edited the first time and a slug gets generated.

**Comment by angrybrad** [comment]
> Posted by Brandon Kelly (brandon@pixelandtonic.com) on 2015/03/06 03:54:45 +0000
-----------------------------------

There are two behaviors that are affecting this:

1. Javascript will only automatically populate the Slug when creating a brand new entry that doesn&#39;t have a Title and Slug yet. (There are a couple reasons for this – to reduce the chances of breaking URLs, and also because the Slug may have actually been deliberately chosen and we wouldn’t want to mess with that.)

2. When you first save a new entry with multiple locales, all of its content (including its Title and Slug) will get copied over to each of its other locales.

So, when you go to edit another locale, by that point it already has its own Title and Slug (copied from the first locale), so any changes made to the Title will not have an effect on the slug.

There is an easy way around this, if you just want Craft to generate a new slug for your new locale: clear out the Slug field’s value. Any time you submit an entry without a Slug, Craft will automatically generate a new one for it when saving the entry.

**Comment by angrybrad** [comment]
> Posted by Erwin Heiser (erwin.heiser@gmail.com) on 2015/03/10 19:49:28 +0000
-----------------------------------

This always felt like a bug or at least an oversight, so yes, totally agree.

**Comment by angrybrad** [comment]
> Posted by Kaspar (bonjour@kasparallenbach.ch) on 2015/03/11 14:14:49 +0000
-----------------------------------

I absolutely agree.

**Comment by angrybrad** [comment]
> Posted by Jérôme Coupé (jercoupe@gmail.com) on 2015/03/12 05:21:11 +0000
-----------------------------------

I couldn&#39;t agree more. Craft&#39;s localisation features are second to none, but this little thing has been a real annoyance, not especially for me, but for the clients that have to maintain sites and post content. They have to remember to change the slug / wipe it the first time they switch locales.

Not sure there is an easy fix (wiping the slug field clean upon changing locales for the first time or upon title change ?) but that would improve the user experience by a massive amount while working on multilingual sites.

The tricky part here is that only needs to be done once, upon switching locales for the first time for any given entry.

**Comment by ursbraem** [comment]
What about another option in the "Save" button: "Save and update slug"?

![image](https://user-images.githubusercontent.com/1446443/72943083-613bd300-3d75-11ea-89ee-ea3c870cff1a.png)

This would be loads more editor friendly than clearing the slug out manually. There could still be a "SEO-Warning" after clicking it to avoid over-usage by unknowing editors.

**Comment by ursbraem** [comment]
That said, it's a workaround... a possible more elegant solution would be to **clear out the slug automatically whenever an entry is opened that hasn't been manually edited once yet**. 

If no changes are made to the title, the same slug as before will be saved so no harm done. 

Wdy think?

**Comment by brandonkelly** [comment]
I don’t think people should have to worry about their entry’s URL possibly changing when they edit its title.

**Comment by ursbraem** [comment]
Sure, but on the other hand, now we have to tell the editors: whenever you translate an article, don't forget to clear out the slug before saving, else you'll have a half-french URL in your italian site.

**Comment by billmn** [comment]
Yeah I agree with @ursbraem, is annoying for editors remember to clear the slug field before save

**Comment by jan-dh** [comment]
Is this something that's on the radar for 4.0?
